### DISCUSSION TITLE
Include .env in results

### DISCUSSION THREAD

**Discussion by joshmedeski** [BurntSushi/ripgrep#2428] [Q&A] [answered]
Hello, I'd like to include my `.env` file in these results:

```
rg --files --hidden
```

`.env` is in my `.gititnore` and I'd like to keep it that way, however I'd like to be able to see it in my results. I'm hoping their is an flag that can be added to tell ripgrep I want to not show all files from gitignore except `.env`.

Something like this?

```
rg --files --hidden --include ".env"
rg --files --hidden --exclude "!.env"
```

I've spent some time in the docs, discussions and issues and can't find what I'm looking for.

Is this possible?

**Comment by BurntSushi** [comment, marked as answer]
Probably the simplest approach here is to do `echo '!/.env' > .rgignore`, which whitelists it. But then it will appear in results even without `--hidden`. But I think that's the best you can do. ripgrep is not arbitrarily flexible in this area.

**Reply by joshmedeski** [reply]
Great, that works on one project. Can I set a global `.rgignore` for all projects?

**Reply by BurntSushi** [reply]
Use `--ignore-file`. You can put that in an alias on ripgrep's config file.

**Comment by steliyan** [comment]
I am a bit at a loss here. I can't really get un-ignoring to work when specifying an ignore file through the CLI. I feel I missing something stupid on my side, but I am not sure what it is.

Here's my setup/project for an easy repro.

## ripgrep setup

Here's my setup (dotfiles):

```
$ echo $RIPGREP_CONFIG_PATH
/Users/steliyan/.config/ripgrep/ripgreprc
```

```
$ ls --tree ~/.config/ripgrep
ripgrep
├── ignore
└── ripgreprc
```

`ignore` contents: 
```
!.env
```

`ripgreprc` contents: 
```
--hidden
```

## Sample project
```
$ ls --tree ~w/rg-sample -A --icon never
rg-sample
├── .env
├── .gitignore
├── app.js
└── hidden-by-gitignore
```

`.gitignore` contents:
```
.env
hidden-by-gitignore
```

## Command
`\rg --hidden --ignore-file=~/.config/ripgrep/ignore --files .`

The output is as follows, note that `.env` is missing:
```
./.gitignore
./app.js
```

If I add `app.js` to the `~/.config/ripgrep/ignore` file then the `app.js` is properly filtered out from the output.

Also, if I move the `~/.config/ripgrep/ignore` as `~/.rgignore` (or `~/.ignore`) then `.env` file shows in the output as expected.

### Debug output

Here's an excerpt from the debug output regarding the `.env`:
```
DEBUG|ignore::walk|crates/ignore/src/walk.rs:1741: ignoring ./.env: Ignore(IgnoreMatch(Gitignore(Glob { from: Some("./.gitignore"), original: ".env", actual: "**/.env", is_whitelist: false, is_only_dir: false })))
```

**Reply by BurntSushi** [reply]
`--ignore-file` has lower precedent than gitignore in your directory tree. In your case, `.env` is being ignored explicitly by a gitignore file rather than being ignored merely because it is a hidden file. So the technique of using `--ignore-file` to whitelist it won't work.

Above all else, whitelisting requires a higher precedent rule than what is causing the file to be ignored. So, as the debug message tells you, if `.env` is being ignored by a rule in `./.gitignore`, then you can add a whitelist rule to `./.rgignore` to un-ignore it.

**Reply by steliyan** [reply]
So, there is no way to override the `.gitignore` behavior with a file outside the current directory (or parent directories)?

**Reply by BurntSushi** [reply]
Correct.

**Reply by steliyan** [reply]
Fair, I am glad I was not losing my mind. :))

Btw, thanks for all the efforts put into `ripgrep`, I am fully aware it's not possible to cover all possible use-cases and feature requests. ;)

**Reply by tomaszferens** [reply]
I somehow managed to achieve this here: https://github.com/LunarVim/LunarVim/discussions/3770#discussioncomment-11648549

1. I have a .gitignore file that ignores ".env*" files in a project
2. I created an `.rgignore` filet at $HOME/.rgignore with the following content: `!.env*`

I can now grep/find  `.env` files :)

**Reply by pinnheads** [reply]
> I somehow managed to achieve this here: [LunarVim/LunarVim#3770 (reply in thread)](https://github.com/LunarVim/LunarVim/discussions/3770#discussioncomment-11648549)
> 
>     1. I have a .gitignore file that ignores ".env*" files in a project
> 
>     2. I created an `.rgignore` filet at $HOME/.rgignore with the following content: `!.env*`
> 
> 
> I can now grep/find `.env` files :)

This worked for me as well! Thanks for sharing
